wrappers/services.go

```diff
@@ -61,18 +61,30 @@ func generateSnapServiceFile(app *snap.AppInfo) ([]byte, error) {
 func stopService(sysd systemd.Systemd, app *snap.AppInfo, inter interacter) error {
 	serviceName := app.ServiceName()
 	tout := serviceStopTimeout(app)
-	if err := sysd.Stop(serviceName, tout); err != nil {
-		if !systemd.IsTimeout(err) {
+
+	stop := func(serviceName string) error {
+		if err := sysd.Stop(serviceName, tout); err != nil {
+			if !systemd.IsTimeout(err) {
+				return err
+			}
+			inter.Notify(fmt.Sprintf("%s refused to stop, killing.", serviceName))
+			// ignore errors for kill; nothing we'd do differently at this point
+			sysd.Kill(serviceName, "TERM")
+			time.Sleep(killWait)
+			sysd.Kill(serviceName, "KILL")
+		}
+
+		return nil
+	}
+
+	for _, socket := range app.Sockets {
+		err := stop(filepath.Base(socket.File()))
+		if err != nil {
```

Comment:
Let's not return earlier if we cannot stop the sockets for whatever reason. Instead:

1. Stop all sockets, collect errors
2. Stop service using fancier logic
3. Return error from service itself, if it failed to stop
4. Return first error in socket stops, if any

Reply:
done